src/cliffordsynthesis/TargetMetricHandler.cpp

```diff
@@ -0,0 +1,142 @@
+    void TargetMetricHandler::makeGateMetric(const SynthesisData& data, bool useMaxSat, bool onlyCNOT) {
+        logicbase::LogicTerm changes = logicbase::LogicTerm(true);
+        // COST
+        if (useMaxSat) {
+            logicbase::LogicTerm cost = logicbase::LogicTerm(0);
+            for (unsigned int gateStep = 1; gateStep < data.timesteps + 1; ++gateStep) {
+                for (unsigned int a = 0; a < data.nqubits; ++a) {
+                    if (!onlyCNOT) {
+                        for (auto gate: Gates::SINGLE_QUBIT_WITHOUT_NOP) {
+                            cost = cost + data.gS[gateStep][Gates::toIndex(gate)][a];
+                        }
+                    }
+                    for (unsigned int b = 0; b <= a; ++b) {
+                        if (a == b) {
+                            continue;
+                        }
+                        cost = cost + data.gTwoQubit[gateStep][a][b] + data.gTwoQubit[gateStep][b][a];
+                    }
+                }
+            }
+            dynamic_cast<logicbase::LogicBlockOptimizer*>(data.lb.get())->minimize(cost);
+        }
+    }
+
+    void TargetMetricHandler::makeDepthMetric(const SynthesisData& data, bool useMaxSat) {
+        using namespace logicbase;
+        // COST
+        if (useMaxSat) {
+            LogicTerm cost = LogicTerm(0);
+            for (unsigned int gateStep = 1; gateStep < data.timesteps + 1; ++gateStep) {
+                LogicTerm anyGate = LogicTerm(true);
+                for (unsigned int a = 0; a < data.nqubits; ++a) {
+                    for (auto gate: Gates::SINGLE_QUBIT_WITHOUT_NOP) {
+                        anyGate = anyGate && !data.gS[gateStep][Gates::toIndex(gate)][a];
+                    }
+                    for (unsigned int b = 0; b <= a; ++b) {
+                        if (a == b) {
+                            continue;
+                        }
+                        anyGate = anyGate && !data.gTwoQubit[gateStep][a][b] && !data.gTwoQubit[gateStep][b][a];
+                    }
+                }
+                cost = cost + LogicTerm::ite(anyGate, LogicTerm(5), LogicTerm(0));
+            }
+            dynamic_cast<LogicBlockOptimizer*>(data.lb.get())->maximize(cost);
+        }
+    }
+
+    void TargetMetricHandler::makeFidelityMetric(const SynthesisData& data, bool useMaxSat, const Architecture& architecture) {
+        if (!architecture.isArchitectureAvailable() || !architecture.isCalibrationDataAvailable()) {
+            util::fatal("No fidelity architecture specified in coupling map.");
+        }
+        // COST
+        if (useMaxSat) {
+            logicbase::LogicTerm cost = logicbase::LogicTerm(0);
+            // For each edge in the coupling map, get the fidelity cost
+            for (const auto& edge: data.reducedCM) {
+                logicbase::LogicTerm fidelity = logicbase::LogicTerm(
+                        (1 - std::log(architecture.getFidelityTable()[edge.first][edge.second])) * 1000);
```

Comment:
The scaling factor `1000` is a magic number and should probably be configurable. Furthermore, is it guaranteed at all that this factor is enough to get a somewhat reasonable cost function if the gate fidelities get high enough?
Say I have a single-qubit gate fidelity of 0.9999 and a two-qubit gate fidelity of 0.999. Then single- and two-qubit gates would be no different under this scaling.